mona_openai/mona_openai.py

```diff
@@ -0,0 +1,243 @@
+from .util.validation_util import validate_and_get_sampling_ratio
+
+EMPTY_DICT = MappingProxyType({})
+
+MONA_ARGS_PREFIX = "MONA_"
+CONTEXT_ID_ARG_NAME = MONA_ARGS_PREFIX + "context_id"
+EXPORT_TIMESTAMP_ARG_NAME = MONA_ARGS_PREFIX + "export_timestamp"
+ADDITIONAL_DATA_ARG_NAME = MONA_ARGS_PREFIX + "additional_data"
+
+
+def _get_monitored_base_class(openai_class):
+    """
+    Returns a class that wrapps the given api class with that
+    api-specific functionality.
+    """
+    class_name = openai_class.__name__
+
+    # TODO(Itai): This is essetially a "switch" statement. We should try to
+    #   use the name to find the exact monitoring-enrichment function and not
+    #   just list all options.
+    if class_name == COMPLETION_CLASS_NAME:
+        return get_completion_class(openai_class)
+    else:
+        raise WrongOpenAIClassException("Class not supported: " + class_name)
+
+
+# TODO(itai): Consider creating some sturct (as NamedTuple or dataclass) for
+#   the specs param.
+
+
+def monitor(openai_class, mona_creds, context_class, specs=EMPTY_DICT):
+    """
+    Returns a Wrapped version of a given OpenAI class with mona
+    monitoring logic.
+    This is the main exposed function of the mona_openai package and
+    probably the only thing you need to use from this package.
+
+    You can use the returned class' "create" and "acreate" functions
+    exactly as you would the original class, and monitoring will be
+    taken care of for you.
+
+    This client will automatically monitor for you things like latency,
+    prompt and response lengths, number of tokens, etc., along with any
+    endpoint parameter usage (e.g., it tracks the "temperature" and
+    "max_tokens" params you use).
+
+    You can also add other named args when calling "create" or
+    "acreate" by using a new named argument called
+    "MONA_additional_data" and set it to any JSON serializable
+    dictionary.
+    This allows you to add metadata about the call such as a prompt
+    template ID, information about the context in which the API call is
+    made, etc...
+
+    Furthermore, you can add to create/acreate functions mona specific
+    arguments:
+        MONA_context_id: The unique id of the context in which the call
+            is made. By using this ID you can export more data to Mona
+            to the same context from other places. If not used, the
+            "id" field of the OpenAI Endpoint's response will be used.
+        MONA_export_timestamp: Can be used to simulate as if the
+            current call was made in a different time, as far as Mona
+            is concerned.
+
+    The returned monitored class also exposes a new class method called
+    "get_mona_clients", which allows you to retrieve the clients and
+    use them to export more data or communicate directly with Mona's
+    API
+
+    Read more about Mona and how to use it in Mona's docs on
+    https://docs.monalabs.io.
```

Comment:
Maybe we should put here a link directly to how export with the client?

Reply:
I think if someone is going to go deeper to Mona in any way, they should start at the beginning and make sure they have a good idea for the basic concepts of Mona.